pennylane/transforms/qfunc_transforms.py

```diff
@@ -0,0 +1,306 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains transforms for registering qfunc transforms."""
+# pylint: disable=too-few-public-methods
+import functools
+import inspect
+
+import pennylane as qml
+
+
+def make_tape(fn):
+    """Returns a function that generates the tape from a quantum function without any
+    operation queuing taking place.
+
+    This is useful when you would like to manipulate or transform
+    the tape created by a quantum function without evaluating it.
+
+    Args:
+        fn (function): the quantum function to generate the tape from
+
+    Returns:
+        function: The returned function takes the same arguments as the quantum
+        function. When called, it returns the generated quantum tape
+        without any queueing occuring.
+
+    **Example**
+
+    Consider the following quantum function:
+
+    .. code-block:: python
+
+        def qfunc(x):
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x, wires=0)
+
+    We can use ``make_tape`` to extract the tape generated by this
+    quantum function, without any of the operations being queued by
+    any existing queuing contexts:
+
+    >>> with qml.tape.QuantumTape() as active_tape:
+    ...     qml.RY(1.0, wires=0)
+    ...     tape = make_tape(qfunc)(0.5)
+    >>> tape.operations
+    [Hadamard(wires=[0]), CNOT(wires=[0, 1]), RX(0.5, wires=[0])]
+
+    Note that the currently recording tape did not queue any of these quantum operations:
+
+    >>> active_tape.operations
+    [RY(1.0, wires=[0])]
+    """
+
+    def wrapper(*args, **kwargs):
+        active_tape = qml.tape.get_active_tape()
+
+        if active_tape is not None:
+            with active_tape.stop_recording(), active_tape.__class__() as tape:
+                fn(*args, **kwargs)
+        else:
+            with qml.tape.QuantumTape() as tape:
+                fn(*args, **kwargs)
+        return tape
+
+    return wrapper
+
+
+class NonQueuingTape(qml.queuing.AnnotatedQueue):
+    """Mixin class that creates a tape that does not queue
+    itself to the current queuing context."""
+
+    def _process_queue(self):
+        super()._process_queue()
+
+        for obj, info in self._queue.items():
+            qml.queuing.QueuingContext.append(obj, **info)
+
+        qml.queuing.QueuingContext.remove(self)
+
+
+class single_tape_transform:
+    """For registering a tape transform that takes a tape and outputs a single new tape.
+
+    Examples of such transforms include circuit compilation.
+
+    Args:
+        transform_fn (function): The function to register as the single tape transform.
+            It can have an arbitrary number of arguments, but the first argument
+            **must** be the input tape.
+
+    **Example**
+
+    A valid single tape transform is a quantum function that satisfies the following:
+
+    - The first argument must be an input tape

```

Comment:
Is this verified somewhere?

Reply:
Unfortunately not. I'm unsure _how_ it can be verified? Unless the user/dev verifies it inside their custom transform

Reply:
All good. I think the docs are sufficiently clear about this in any case :+1: